tracer/src/Datadog.Trace/Agent/DiscoveryService/DiscoveryService.cs

```diff
@@ -252,9 +281,44 @@ int GetNextSleepDuration(int? previousDuration) =>
                 previousDuration is null ? _initialRetryDelayMs : Math.Min(previousDuration.Value * 2, _maxRetryDelayMs);
         }
 
+        [TestingAndPrivateOnly]
+        internal bool RequireRefresh(string? currentHash, DateTimeOffset utcNow)
```

Comment:
Could there be a race condition here?
Thread A reads  _agentConfigStateHash 
Thread B calls SetCurrentConfigStateHash() and updates hash and time
Thread A reads _agentConfigStateHashUnixTime and gets a fresh timestamp, return false
Not critical though, it would be rare and the consecuences are not severe. WDYT?

Reply:
> Could there be a race condition here?

Yup, absolutely, there could. When I first wrote this, I worked around that by creating a `record State(string Hash, DateTimeOffset utcNow)` and then atomically switching that. The problem is that `SetCurrentConfigStateHash()` is going to be called a lot, so it would result in _either_ a more complex "atomically read, compare, conditionally recreate  , atomically write" comparison, _or_ if we just always write (like we're doing currently) then we end up allocating every time. 

My conclusion was pretty much the same as yours, it should be rare, and not severe (we will pick it up next time), and given this is primarily meant to be a perf improvement, I think it's worth it 🙂